mesonbuild/dependencies/base.py

```diff
@@ -1983,40 +1984,80 @@ def get_partial_dependency(self, *, compile_args=False, link_args=False,
 
 
 class ExtraFrameworkDependency(ExternalDependency):
-    def __init__(self, name, required, path, env, lang, kwargs):
+    system_framework_paths = ('/Library/Frameworks', '/System/Library/Frameworks')
```

Comment:
Why hardcode these and not just use the defaults?

Reply:
See below, we need to know what path was selected to set `self.framework_path`.

Reply:
Also needed for the include path, since setting `-Farg` is not enough with most projects.

Reply:
Can you not query the compiler for those? IMO hardcoding such things just leads to a lot more trouble in the future and meson already unfortunately hardcodes a huge amount of paths…

And can you maybe elaborate why exactly you need those? I do not fully understand the reason and what you mean by "below".

Reply:
Ah ok I get why you need those, still, can you not query the compiler for those? The current approach would fail if I would specify additional Framework search paths with cflags, no?

Reply:
Well, when I said "see below", I meant there's a comment a few lines later that explains why we need the full path:

> We need to know the exact framework path because it's used by the Qt5 dependency class. We also want to avoid searching in an invalid framework path which wastes time and can cause a false positive.

We also need it for the include path, since most cross-platform projects do not use a "framework include", and hence need `-I/path/to/app.framework/Headers`. I'll add a comment about this.

As far as I know, there is no way to query the default framework search path from either [the linker](https://www.unix.com/man-page/osx/1/ld/) or [the compiler](https://clang.llvm.org/docs/ClangCommandLineReference.html). Believe me, I looked 😄. There is also no way to override the default search paths for frameworks.

Either way, hardcoding these paths is not a regression since that's what we did before anyway. If someone finds a case in which this is broken or insufficient, we can look always improve it further then.

Reply:
Apparently this works:

```
echo | clang -F/usr/local/opt/ -v -E -x c -
Apple LLVM version 10.0.0 (clang-1000.11.45.5)
Target: x86_64-apple-darwin18.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple x86_64-apple-macosx10.14.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -E -disable-free -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fno-strict-return -masm-verbose -munwind-tables -target-cpu penryn -dwarf-column-info -debugger-tuning=lldb -target-linker-version 409.12 -v -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/10.0.0 -F/usr/local/opt/ -fdebug-compilation-dir /Users/epirat/Projects/VideoLAN/VLC/build -ferror-limit 19 -fmessage-length 128 -stack-protector 1 -fblocks -fencode-extended-block-signature -fobjc-runtime=macosx-10.14.0 -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o - -x c -
clang -cc1 version 10.0.0 (clang-1000.11.45.5) default target x86_64-apple-darwin18.5.0
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/opt (framework directory)
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/10.0.0/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
# 1 "<stdin>"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 361 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "<stdin>" 2
```

Reply:
That gives a different output for me:

```
$ echo | clang -F/usr/local/opt/ -v -E -
Apple LLVM version 10.0.0 (clang-1000.11.45.5)
Target: x86_64-apple-darwin18.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple x86_64-apple-macosx10.14.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -E -disable-free -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fno-strict-return -masm-verbose -munwind-tables -target-cpu penryn -dwarf-column-info -debugger-tuning=lldb -target-linker-version 409.12 -v -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/10.0.0 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -F/usr/local/opt/ -I/usr/local/include -fdebug-compilation-dir /Users/nirbheek/projects/meson/meson.git -ferror-limit 19 -fmessage-length 363 -stack-protector 1 -fblocks -fencode-extended-block-signature -fobjc-runtime=macosx-10.14.0 -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o - -x c -
clang -cc1 version 10.0.0 (clang-1000.11.45.5) default target x86_64-apple-darwin18.0.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/opt (framework directory)
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/10.0.0/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks (framework directory)
End of search list.
# 1 "<stdin>"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 361 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "<stdin>" 2

```

Why doesn't that include `/Library/Frameworks`? Very suspicious.

Reply:
Ah, that changes based on the value of `-isysroot` being passed to clang. In your case there's no default value set. In my case it's set to `-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk`. If I pass `-isysroot /` I get your values.

Reply:
Ah yes, that seems like thats how it should be though, no?
When compiling with a specific SDK you usually want/expect it to use its framework stubs instead of the system frameworks.

Reply:
Yes, indeed.

Reply:
Updated, please review again!